dtls.c

```diff
@@ -4061,15 +4062,30 @@ dtls_handle_message(dtls_context_t *ctx,
       dtls_security_parameters_t *security = dtls_security_params_epoch(peer, dtls_get_epoch(header));
       if (!security) {
         dtls_alert("No security context for epoch: %i\n", dtls_get_epoch(header));
-        data_length = -1;
+        return -1;
       } else {
+        uint16_t msg_epoch = dtls_get_epoch(header);
+
+        if (peer->handshake_params &&
+            peer->handshake_params->hs_epoch != msg_epoch &&
+            peer->state != DTLS_STATE_WAIT_FINISHED) {
+          if ((uint16_t)(peer->handshake_params->hs_epoch + 1) == msg_epoch) {
+            dtls_info("Next epoch packet arrived before handshake complete\n");
+            /* Just validate the packet, reporting any issues */
+            decrypt_verify(peer, msg, rlen, &data);
```

Comment:
I am not sure if this extra operation is a good idea. Why do you think it is necessary?

Reply:
The decrypt_verify() would otherwise not take place on finished in the new epoch following receipt of change cipher spec and throw up any errors as this packet is not buffered.  I agree that subsequent decryption of application packets would fail if there is an error here, but this may not be correctly detected at this point.

Reply:
The `decrypt_verify()` essentially just removes the "other security parameter" slot that is used internally to store values from the last epoch (it is a noop if there are none or if values for the next epoch are available). I still wonder why it would make a difference whether this is done here or in the next call to `decrypt_verify()`.

Reply:
Using the dtls-fuzzer from Issue #79, if decrypt_verify() @4075 is removed, the following application data is still decrypted correctly.

My only concern with current fix is if this decrypt_verify() is removed, there never will be a validation of the out of order Finish packet.